acurl/src/acurl.pyx

```diff
@@ -21,7 +21,7 @@ class AcurlError(Exception):
 # with the python asyncio library we need to call back into python.  So these
 # functions do take the GIL, as indicated by `with gil` in their signatures.
 # The performance impact of this is yet tbd.
-cdef int handle_socket(CURL *easy, curl_socket_t sock, int action, void *userp, void *socketp) with gil:
+cdef int handle_socket(CURL *easy, curl_socket_t sock, int action, void *userp, void *socketp) noexcept with gil:
```

Comment:
I think it might help to have some commentary about why we think noexcept is correct for these functions.  For instance, this one has code that explicitly raises an exception (although it's expected never to be hit...)  The other functions that call python code might raise in subtler ways.

I understand why we are doing this (I think) -- because we need to use these fns as C callbacks (in a context where exceptions won't get handled anyway).  But we can help spell that out for future readers of this code, especially those who might be confused seeing `raise` in a `noexcept` function, and not immediately be able to call to mind the relevant facts about the usage of callbacks etc.

WDYT?

Reply:
We could panic rather than raising an exception, I suppose...the only way we'd hit that branch of code is if there was a breaking change to the curl API, or there was some kind of data corruption.  In either case I think we don't want it to be silent.  (I guess til now the exception would have been ignored, so it was a false sense of safety).

> is the second or action == CURL_POLL_INOUT in this function is redundant

No, since if the action is `INOUT` we need to add both the reader and the writer to the event loop.  There are two (actually three) separate `if`s, it's not an `elif`.  I guess we could refactor to use one of the fancy new `match` statements (but that would lead to code duplication, since we'd need the `add_reader` and `add_writer` calls twice, once in the separate `IN` or `OUT` case and once in the combined `INOUT` case).